### ISSUE TITLE
tttt

### ISSUE THREAD

**Issue by Censwin** [Censwin/myblog#3]
// ==UserScript==
// @name        STU
// @namespace   Violentmonkey Scripts
// @match        *://stu.5zk.com.cn/zk8exam/jp_wiki_study.php?*
// @match        *://stu.5zk.com.cn/zk8exam/login.php
// @match        *://stu.5zk.com.cn/zk8exam/mycourse.php
// @grant       none
// @version     1.0
// @author      -
// @description 3/13/2023, 9:51:55 PM
// ==/UserScript==
(function () {
    window.__unFinishedList = undefined;
    window.__unFinishedIndex = undefined;
    window.onload = function () {
        if (window.location.href === "https://stu.5zk.com.cn/zk8exam/login.php") {
            $("[name=admin_name]").val("235310534050");
            $("#admin_password").val("02011X");
            setTimeout(function () {
                $("form").submit();
            }, 1000)
            return;
        }
        if (window.location.href === "https://stu.5zk.com.cn/zk8exam/mycourse.php") {
            locateLessonPage();
            return;
        }
        if (!getPlayList()) return;
        mutedVideoAndPlay();
        banAlter();
        submitAnswer();
        player.on('ended', function (ended) {
            onplayend2()
        });
    }
    /*
     * description: by the chrome browser restriction, muted video first then play
     * */
    function mutedVideoAndPlay() {
        document.getElementById("player-container-id_html5_api").setAttribute("muted", "")
        player.volume(0);
        setTimeout(function () {
            player.play();
        })
    }
    /*
     * description: get the lessons list
     * */
    function getPlayList() {
        const __classUL = [...document.getElementsByClassName('block-content')].filter(item => {
            return item.children[0].tagName == "UL"
        })[0];
        const __AllLessonArray = [...__classUL.getElementsByTagName("a")];
        __unFinishedList = __AllLessonArray.filter(item => {
            const value = item.children[0].attributes.style.value;
            return value === 'color:#FF8000;' || value === 'color:#EA0000;';
        })
        if (__unFinishedList.length === 0) return false;
        __unFinishedIndex = __unFinishedList.findIndex(item => item.href == window.location.href);
        if (__unFinishedIndex === -1) {
            locateLastItem(__unFinishedList)
        }
        const index = __AllLessonArray.findIndex(item => item.href == window.location.href);
        showNotic(__AllLessonArray, index);
        return true;
    }
    /*
     * description: block the alert
     * */
    function banAlter() {
        alert = function (v) {
            if (v === "您今天已经新学了10个课时（450分钟），请明天再来学习新课程！为保证学习效果，每天最多新学10个课时。") {
                player.pause();
                setMark();
                // confirm(v);
                window.location = "https://stu.5zk.com.cn/zk8exam/mycourse.php"
            } else {
                player.play();
            }
        };
    }
    // append some information on top
    function showNotic(list, index) {
        const notic = document.createElement("div");
        notic.classList.add("content");
        const next = __unFinishedList[__unFinishedIndex + 1] ? __unFinishedList[__unFinishedIndex + 1].innerText :
            "完成";
        notic.innerHTML =
            `<h2>课程总共 ${list.length} 节，当前第 ${index+1} 节，下一节：${next}</h2><p>课程播完后、当天超过10小时 将自动返回课程列表</p>`;
        notic.style.setProperty("color", "red");
        const container = document.getElementById("main-container");
        container.insertBefore(notic, container.firstChild);
    }
    /*
     * description: auto submit answer, not sure the backend whether to save something about the exam during the video playing,
     * if not, we could just set the key isnot_st to 0.
     * */
    function submitAnswer() {
        if (document.getElementById("isnot_st").value == "1") {
            setTimeout(() => {
                ktcs();
            }, 2000);
            setTimeout(function () {
                const target = document.getElementById("layui-layer1").getElementsByTagName("iframe")[0];
                if (target) {
                    const radioList = [...target.contentWindow.document.getElementsByTagName("label")];
                    const radioListWithoutTitle = radioList.slice(1);
                    const randomNum = randomIntBetween(0, radioListWithoutTitle.length - 1);
                    radioListWithoutTitle[randomNum].click();
                    const submitBtn = target.contentWindow.document.getElementsByTagName("button")[0];
                    setTimeout(() => {
                        submitBtn.click();
                    }, 3000)
                    setTimeout(() => {
                        const [completeBtn] = [...document.getElementsByClassName(
                            "layui-layer-btn0")].filter(item => item.textContent === '我已完成答题')
                        completeBtn.click();
                    }, 5000)
                }
                document.getElementById("isnot_st").value = "0";
            }, 5 * 1000)
        }
    }

    /*
     * description: return a random number for the exam;
     * */
    function randomIntBetween(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    // locateLastItem
    function locateLastItem(playList) {
        window.location = __unFinishedList[0].href;
    }
    /*
     * description: if the study time was longer than 10 hours Today, set a key to avoid catalog page auto relocate
     * */
    function setMark() {
        const today = new Date().toLocaleDateString();
        localStorage.setItem('excessDate', today);
    }

    function locateLessonPage() {
        const scoreTargetArray = $('tbody tr td div.row:nth-child(3) div.col-3:nth-child(2) nobr');
        const index = [...scoreTargetArray].findIndex(item => item.innerText < 100);
        const lessonsTargetArray = $('tbody tr td:nth-child(2) nobr');
        $("#main-container div.content p").eq(0).after(
            `<h3 id="__located" style="color:red;">已定位 ${lessonsTargetArray[index].innerText}</h3>`);
        $("#__located").after(`<b id="__countdown" style="color:red;"></b>`)
        if (!isExcess()) {
            let i = 5;
            let intervalId = setInterval(function () {
                if (i == 0) {
                    clearInterval(intervalId);
                    lessonsTargetArray[index].children[0].click();
                } else {
                    $("#__countdown").text(`准备跳转 ${i}`);
                    i--;
                }
            }, 1000);
        }
    }

    function isExcess() {
        const today = new Date().toLocaleDateString();
        const lastDate = localStorage.getItem('excessDate');
        if (lastDate === today) {
            $("#__located").after(`<h3 style="color:red;">今日学习时间已超时</h3>`)
            setTimeout(() => {
              window.reload();
            }, 1000 * 60 * 60 * 6);
            return true;
        } else {
            if (lastDate) {
                localStorage.removeItem('excessDate');
            }
            return false;
        }
    }

    function onplayend2() {
        var jjwikiid = document.getElementById("jjwikiid").value;
        var TypeId = document.getElementById("TypeId").value;
        var cpidlist = document.getElementById("cpidlist").value;
        var cur_jindu = 999999;
        var userlog_id = document.getElementById("userlog_id").value;
        var sp_type = document.getElementById("sp_type").value;
        var ajax = new Ajax();
        ajax.getInfo("learn_jjsp.action.php", "GET", "app", "&jjwikiid=" + jjwikiid + "&TypeId=" + TypeId +
            "&cur_jindu=" + cur_jindu + "&userlog_id=" + userlog_id,
            function (c) {
                if (c == '888888') {
                    player.pause();
                    alert('学习进度异常，请重新开始学习！！');
                    window.location.reload();
                } else {
                    document.getElementById("max_jindu").value = c;
                    document.getElementById('mj_m_01').innerHTML = parseInt(c / 60);
                    document.getElementById('mj_m_02').innerHTML = parseInt(c / 60);

                    document.getElementById("cur_jindu").value = cur_jindu;
                    //章节练习
                    var is_zy = document.getElementById("is_zy").value;
                    var isnot_st = document.getElementById("isnot_st").value;
                    if (is_zy == 0) {
                        setTimeout(function () {
                            if (__unFinishedList[__unFinishedIndex + 1]) {
                                window.location = __unFinishedList[__unFinishedIndex + 1].href
                            } else {
                                window.location = "https://stu.5zk.com.cn/zk8exam/mycourse.php"
                            }
                        }, 2000)
                    }
                    return;
                }
            });
    }

})()

**Comment by Censwin** [comment]
```
// ==UserScript==
// @name        STU
// @namespace   Violentmonkey Scripts
// @match        *://stu.5zk.com.cn/zk8exam/jp_wiki_study.php?*
// @match        *://stu.5zk.com.cn/zk8exam/login.php
// @match        *://stu.5zk.com.cn/zk8exam/mycourse.php
// @grant       none
// @version     1.0
// @author      -
// @description 3/13/2023, 9:51:55 PM
// ==/UserScript==
(function () {
    window.__unFinishedList = undefined;
    window.__unFinishedIndex = undefined;
    window.onload = function () {
        if (window.location.href === "https://stu.5zk.com.cn/zk8exam/login.php") {
            $("[name=admin_name]").val("235310534050");
            $("#admin_password").val("02011X");
            setTimeout(function () {
                $("form").submit();
            }, 1000)
            return;
        }
        if (window.location.href === "https://stu.5zk.com.cn/zk8exam/mycourse.php") {
            locateLessonPage();
            return;
        }
        if (!getPlayList()) return;
        mutedVideoAndPlay();
        banAlter();
        submitAnswer();
        player.on('ended', function (ended) {
            onplayend2()
        });
    }
    /*
     * description: by the chrome browser restriction, muted video first then play
     * */
    function mutedVideoAndPlay() {
        document.getElementById("player-container-id_html5_api").setAttribute("muted", "")
        player.volume(0);
        setTimeout(function () {
            player.play();
        })
    }
    /*
     * description: get the lessons list
     * */
    function getPlayList() {
        const __classUL = [...document.getElementsByClassName('block-content')].filter(item => {
            return item.children[0].tagName == "UL"
        })[0];
        const __AllLessonArray = [...__classUL.getElementsByTagName("a")];
        __unFinishedList = __AllLessonArray.filter(item => {
            const value = item.children[0].attributes.style.value;
            return value === 'color:#FF8000;' || value === 'color:#EA0000;';
        })
        if (__unFinishedList.length === 0) return false;
        __unFinishedIndex = __unFinishedList.findIndex(item => item.href == window.location.href);
        if (__unFinishedIndex === -1) {
            locateLastItem(__unFinishedList)
        }
        const index = __AllLessonArray.findIndex(item => item.href == window.location.href);
        showNotic(__AllLessonArray, index);
        return true;
    }
    /*
     * description: block the alert
     * */
    function banAlter() {
        alert = function (v) {
            if (v === "您今天已经新学了10个课时（450分钟），请明天再来学习新课程！为保证学习效果，每天最多新学10个课时。") {
                player.pause();
                setMark();
                // confirm(v);
                window.location = "https://stu.5zk.com.cn/zk8exam/mycourse.php"
            } else {
                player.play();
            }
        };
    }
    // append some information on top
    function showNotic(list, index) {
        const notic = document.createElement("div");
        notic.classList.add("content");
        const next = __unFinishedList[__unFinishedIndex + 1] ? __unFinishedList[__unFinishedIndex + 1].innerText :
            "完成";
        notic.innerHTML =
            `<h2>课程总共 ${list.length} 节，当前第 ${index+1} 节，下一节：${next}</h2><p>课程播完后、当天超过10小时 将自动返回课程列表</p>`;
        notic.style.setProperty("color", "red");
        const container = document.getElementById("main-container");
        container.insertBefore(notic, container.firstChild);
    }
    /*
     * description: auto submit answer, not sure the backend whether to save something about the exam during the video playing,
     * if not, we could just set the key isnot_st to 0.
     * */
    function submitAnswer() {
        if (document.getElementById("isnot_st").value == "1") {
            setTimeout(() => {
                ktcs();
            }, 2000);
            setTimeout(function () {
                const target = document.getElementById("layui-layer1").getElementsByTagName("iframe")[0];
                if (target) {
                    const radioList = [...target.contentWindow.document.getElementsByTagName("label")];
                    const radioListWithoutTitle = radioList.slice(1);
                    const randomNum = randomIntBetween(0, radioListWithoutTitle.length - 1);
                    radioListWithoutTitle[randomNum].click();
                    const submitBtn = target.contentWindow.document.getElementsByTagName("button")[0];
                    setTimeout(() => {
                        submitBtn.click();
                    }, 3000)
                    setTimeout(() => {
                        const [completeBtn] = [...document.getElementsByClassName(
                            "layui-layer-btn0")].filter(item => item.textContent === '我已完成答题')
                        completeBtn.click();
                    }, 5000)
                }
                document.getElementById("isnot_st").value = "0";
            }, 5 * 1000)
        }
    }

    /*
     * description: return a random number for the exam;
     * */
    function randomIntBetween(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    // locateLastItem
    function locateLastItem(playList) {
        window.location = __unFinishedList[0].href;
    }
    /*
     * description: if the study time was longer than 10 hours Today, set a key to avoid catalog page auto relocate
     * */
    function setMark() {
        const today = new Date().toLocaleDateString();
        localStorage.setItem('excessDate', today);
    }

    function locateLessonPage() {
        const scoreTargetArray = $('tbody tr td div.row:nth-child(3) div.col-3:nth-child(2) nobr');
        const index = [...scoreTargetArray].findIndex(item => item.innerText < 100);
        const lessonsTargetArray = $('tbody tr td:nth-child(2) nobr');
        $("#main-container div.content p").eq(0).after(
            `<h3 id="__located" style="color:red;">已定位 ${lessonsTargetArray[index].innerText}</h3>`);
        $("#__located").after(`<b id="__countdown" style="color:red;"></b>`)
        if (!isExcess()) {
            let i = 5;
            let intervalId = setInterval(function () {
                if (i == 0) {
                    clearInterval(intervalId);
                    lessonsTargetArray[index].children[0].click();
                } else {
                    $("#__countdown").text(`准备跳转 ${i}`);
                    i--;
                }
            }, 1000);
        }
    }

    function isExcess() {
        const today = new Date().toLocaleDateString();
        const lastDate = localStorage.getItem('excessDate');
        if (lastDate === today) {
            $("#__located").after(`<h3 style="color:red;">今日学习时间已超时</h3>`)
            setTimeout(() => {
              window.reload();
            }, 1000 * 60 * 60 * 6);
            return true;
        } else {
            if (lastDate) {
                localStorage.removeItem('excessDate');
            }
            return false;
        }
    }

    function onplayend2() {
        var jjwikiid = document.getElementById("jjwikiid").value;
        var TypeId = document.getElementById("TypeId").value;
        var cpidlist = document.getElementById("cpidlist").value;
        var cur_jindu = 999999;
        var userlog_id = document.getElementById("userlog_id").value;
        var sp_type = document.getElementById("sp_type").value;
        var ajax = new Ajax();
        ajax.getInfo("learn_jjsp.action.php", "GET", "app", "&jjwikiid=" + jjwikiid + "&TypeId=" + TypeId +
            "&cur_jindu=" + cur_jindu + "&userlog_id=" + userlog_id,
            function (c) {
                if (c == '888888') {
                    player.pause();
                    alert('学习进度异常，请重新开始学习！！');
                    window.location.reload();
                } else {
                    document.getElementById("max_jindu").value = c;
                    document.getElementById('mj_m_01').innerHTML = parseInt(c / 60);
                    document.getElementById('mj_m_02').innerHTML = parseInt(c / 60);

                    document.getElementById("cur_jindu").value = cur_jindu;
                    //章节练习
                    var is_zy = document.getElementById("is_zy").value;
                    var isnot_st = document.getElementById("isnot_st").value;
                    if (is_zy == 0) {
                        setTimeout(function () {
                            if (__unFinishedList[__unFinishedIndex + 1]) {
                                window.location = __unFinishedList[__unFinishedIndex + 1].href
                            } else {
                                window.location = "https://stu.5zk.com.cn/zk8exam/mycourse.php"
                            }
                        }, 2000)
                    }
                    return;
                }
            });
    }

})()
```
